### DISCUSSION TITLE
Run job only if folder changed

### DISCUSSION THREAD

**Discussion by nahtnam** [community/community#25669] [Actions]
<p>Hello,</p>
<p>I currently have a job as part of my deploy workflow which connects to my db, runs migrations and seeds the database. That takes about 2 minutes total. I was wondering if there was any way to only this job or those commands IF the <code>db</code> folder changed?</p>
<p>I know I can write a bash script and use some sort of git command to check but I was wondering if there is an in-built way to do this?</p>
<p>Thanks!</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/nahtnam">@nahtnam</a>,</p>
<p>There is not a built-in way to do that. You need to use some commands or actions to list all the modified files and check if all of them are in the <strong>db</strong> folder.<br>
Suppose the <strong>db</strong> folder is located at the root of the repository, you can reference to the example below:</p>
<pre><code class="lang-yaml">jobs:
  check:
    name: Check files
    outputs:
      run_job: ${{ steps.check_files.outputs.run_job }}
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2
        with:
          fetch-depth: 2

      - name: check modified files
        id: check_files
        run: |
          echo "=============== list modified files ==============="
          git diff --name-only HEAD^ HEAD
          
          echo "========== check paths of modified files =========="
          git diff --name-only HEAD^ HEAD &gt; files.txt
          while IFS= read -r file
          do
            echo $file
            if [[ $file != db/* ]]; then
              echo "This modified file is not under the 'db' folder."
              echo "::set-output name=run_job::false"
              break
            else
              echo "::set-output name=run_job::true"
            fi
          done &lt; files.txt

  job_for_db:
    name: Job for 'db' folder
    needs: check
    if: needs.check.outputs.run_job == 'true'
    runs-on: ubuntu-latest
    steps:
      - name: run for 'db' folder
        run: echo "This job runs only when all the modified files are under the 'db' folder."
</code></pre>
<p>The following is a similar case as reference:</p><aside class="quote quote-modified" data-post="6" data-topic="117344">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/how-to-execute-jobs-if-folder-updated-recursive/117344/6">How to execute Jobs if folder updated (recursive)?</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    <a class="mention" href="/heathsnow">@heathsnow</a>, 
You can use “git diff” command to list all the changed files, and then check if all the changed files are under the directory “subdir/env1”. If all the changed files are under that directory, set an output with the value is true, otherwise false. Then according to the value of the output to run or not run the subsequent job which has the step  echo “Updates made to env1”. 
A simple example: 
jobs:
  check:
    name: Check changed files
    outputs:
      run_job: ${{ steps.check_fil…
  </blockquote>
</aside>

**Reply by markoueis** [reply]
What if you push multiple commits, this will only detect the changes in the last commit?

**Reply by stejackson94** [reply]
This was quite helpful for a problem i was just stuck on, thank you

**Comment by nahtnam** [comment]
<p>Thank you! That did the trick, however I’m stuck on another issue. I want my next job (deploy) to wait for <code>database</code> to finish before running, but if <code>database</code> is skipped then run it immediately. Any idea how to do that?</p>

**Comment by nahtnam** [comment]
<p>I found a “hack”. I just moved the <code>if</code> statement you provided to every command inside of the <code>database</code> job instead of at the job level if that makes sense</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/nahtnam">@nahtnam</a>,</p>
<p>I recommend you to use <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds" rel="nofollow noopener"><strong>needs</strong></a> and <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions" rel="nofollow noopener"><strong>Job status check functions</strong></a> to do that.<br>
In your case, you can try like as below:</p>
<pre><code class="lang-yaml">jobs:
  . . .

  database:
    name: job database
    . . .

  deploy:
    name: job deploy
    needs: database
    if: always()
    . . .
</code></pre>
<p>In this way:</p>
<ul>
<li>If job <strong>database</strong> is executed, job <strong>deploy</strong> will run after <strong>database</strong> completes, regardless of whether <strong>database</strong> is success or failure.</li>
<li>If job <strong>database</strong> is skipped, job <strong>deploy</strong> also will run.</li>
</ul>

**Comment by nahtnam** [comment]
<p>That worked, thank you so much! Just wondering, where did you learn all of this? The docs are kind of hard to sort through, so I was wondering if there was another guide or something</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/nahtnam">@nahtnam</a>,</p>
<p>In fact, the official documentation has provided a quite detailed guidance of these knowledge and technologies. The appropriate engineering team also will update the documentation in time for the new features, or improve some descriptions in the documentation so that readers can better understand the relevant knowledge.</p>
<p>You can pay more attention to these documents, and learn almost all the technical guidance about GitHub Actions.<br>
Of course, if you have any questions or troubles on GitHub Actions, please feel free to contact us. We will provide you with help and support as mush as we can.</p>

**Comment by nahtnam** [comment]
<p>Thank you, I think its just a lot of information all in one place so its a little overwhelming. Also I am facing an issue.</p>
<p>So I made a change to my <code>prisma</code> folder (database) and for some reason my database job was still skipped:</p>
<pre><code class="lang-auto">check_database:
    outputs:
      run_job: ${{ steps.check_files.outputs.run_job }}
    runs-on: ubuntu-latest
    needs: test
    if: github.ref == 'refs/heads/master'
    steps:
      - name: Checkout code
        uses: actions/checkout@v2
        with:
          fetch-depth: 2

      - name: check modified files
        id: check_files
        run: |
          echo "=============== list modified files ==============="
          git diff --name-only HEAD^ HEAD

          echo "========== check paths of modified files =========="
          git diff --name-only HEAD^ HEAD &gt; files.txt
          while IFS= read -r file
          do
            echo $file
            if [[ $file = prisma/* ]]; then
              echo "This modified file is under the 'prisma' folder."
              echo "::set-output name=run_job::true"
              break
            else
              echo "::set-output name=run_job::false"
            fi
          done &lt; files.txt

  database:
    runs-on: ubuntu-latest
    needs: check_database
    if: needs.check.outputs.run_job == 'true' &amp;&amp; github.ref == 'refs/heads/master'
    env:
      DATABASE_URL: ${{ secrets.PRODUCTION_DATABASE_URL }}
    steps:
      - uses: actions/checkout@v2
      - name: Use Node.js 12.x
        uses: actions/setup-node@v1
        with:
          node-version: '12.x'
      - run: npm ci
      - run: npx prisma generate
      - run: npx prisma migrate up --experimental
      - run: npx ts-node prisma/_seeds/production/index.ts
</code></pre>
<p>In the <code>check_database</code> job, I see this output</p>
<pre><code class="lang-auto">=============== list modified files ===============
// bunch of files here
========== check paths of modified files ==========
prisma/_seeds/development/index.ts
This modified file is under the 'prisma' folder.
</code></pre>
<p>so it looks like it is recognizing the change but for some reason the database job is still skipped.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/nahtnam">@nahtnam</a>,</p>
<p>You are referencing to an inconsistent <strong>job id</strong> in the <strong>if</strong> conditional, the expression syntax is “<strong>needs.&lt;job id&gt;.outputs.&lt;output name&gt;</strong>”.</p>
<pre><code class="lang-yaml">needs: check_database
if: needs.check.outputs.run_job == 'true' &amp;&amp; ...
</code></pre>
<p>it should be “<strong>check_database</strong>”, not “<strong>check</strong>”, you do not set job “<strong>database</strong>” needs any one job which named “<strong>check</strong>”, and there is no job “<strong>check</strong>” in your workflow.<br>
So, please correct your workflow like as below, and try again:</p>
<pre><code class="lang-yaml">needs: check_database
if: needs.check_database.outputs.run_job == 'true' &amp;&amp; ...
</code></pre>

**Comment by nahtnam** [comment]
<p>My mistake, thank you so much!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/nahtnam">@nahtnam</a>,<br>
You’re welcome. And glad that my suggestions can help you solve the problems.<br>
If you have any other questions about this ticket, feel free to tell me.</p>

**Comment by tbhaxor** [comment]
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="118292">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p><code>::set-output name=run_job::false</code></p>
</blockquote>
</aside>
<p>Can you please explain this segment ???</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/tbhaxor">@tbhaxor</a>,</p>
<blockquote>
<p>::set-output name=run_job::false</p>
</blockquote>
<p>This is setting an output of a step and can be used in the subsequent steps.</p>
<p>To view more details, you can see:</p>
<ul>
<li><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter" rel="noopener nofollow ugc"><strong>Setting an output parameter</strong></a></li>
<li><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs" rel="noopener nofollow ugc"><strong>jobs.&lt;jobs_id&gt;.outputs</strong></a></li>
</ul>

**Comment by alpha-potato** [comment]
<p>Can try use ‘paths’</p>
<pre><code class="lang-auto">name: path_trigger

on:
  push:
    branches: [ main ]
    paths:
      - some_random_path/**
      - requirement.txt

jobs:
  deploy:
    runs-on: [ ubuntu-latest ]
    steps:
      - uses: actions/checkout@v2
      - name: deploy script
        run: .github/workflows/scripts/your_job.sh
</code></pre>

**Reply by LucBerge** [reply]
Using `paths` will apply it to all the jobs in the workflow, not a specific one.

**Comment by PaulRBerg** [comment]
<p>There seems to be an off-the-shelf GitHub action that solves this: <a href="https://github.com/dorny/paths-filter" rel="noopener nofollow ugc">paths-filter</a>. Example from their README:</p>
<pre><code class="lang-yaml">- uses: dorny/paths-filter@v2
  id: changes
  with:
    filters: |
      src:
        - 'src/**'

  # run only if some file in 'src' folder was changed
- if: steps.changes.outputs.src == 'true'
  run: ...
</code></pre>

**Reply by LucBerge** [reply]
It works, thanks

**Reply by brunomgalmeida** [reply]
this worked for me too, thanks

**Comment by Filip13131** [comment]
You Can just use multiple .yml Files in .github/workflows 
then specify On: virable as you need it

**Reply by sdarwin** [reply]
Example:

```
on:
  pull_request:
    types:
      - opened
    branches:
      - 'releases/**'
    paths:
      - '**.js'
```

**Reply by alobaton-pricesmart** [reply]
It worked for me. I removed the `types: - opened` filter and separated the workflows into different files per each directory. Thanks!

**Comment by LucBerge** [comment]
I'm trying to do it like this but I cannot make it work:

```yaml
  Flow_A_:
    runs-on: ubuntu-latest
    if: contains(github.event.commits[0].modified[*], 'data/')

    steps:
      - name: Echo A
        run: |
          echo "This is Flow A"
```

The return value of the `contains(...)` function is always `false` and there is no way to list the array to check the values.

As for now, using @paulrberg solution is the best way to do it

**Reply by piotrekkr** [reply]
@LucBerge Commit may contain multiple files and `github.event.commits[0].modified[*]` will be an array of strings (my guess here). `contains()` will search in this array by doing exact comparison of each element in array to a string. This is probably why it does not work.

Have you tried to just display what is inside this `github.event.commits[0].modified[*]`? Something like 
```
echo '${{ toJSON(github.event.commits[0].modified[*]) }}' | jq
```
just to be sure if inside this array there is a string `data/` (not substring).

**Comment by kyledrake6o9** [comment]
Rose conduct car, film, grit flavour behind spot crazy myself imitate mandate, 10 Fitness knock crew bronze, olive depart, initial real eight one frown
